### PR TITLE
[CARBONDATA-1972][PARTITION] Fix compaction after update of partition table.

### PR DESCRIPTION
**Pull request by ravipesala**
When updation happens on whole data then all old segments needs to be marked as delete. But it is not happening in case of partition table. This PR fixes it.

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [X] Any interfaces changed?
  NO
 - [X] Any backward compatibility impacted?
   NO
 - [X] Document update required?
   NO
 - [X] Testing done
       Tests added
       
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2497/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2657/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1272/

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java [hunk 1]
```diff
@@ -98,12 +96,25 @@ public CarbonOutputCommitter(Path outputPath, TaskAttemptContext context) throws
...
+          context.getConfiguration().get(CarbonTableOutputFormat.SEGMENTS_TO_BE_DELETED, null);
       if (updateTime != null) {
         Set<String> segmentSet = new HashSet<>(
             new SegmentStatusManager(carbonTable.getAbsoluteTableIdentifier())
                 .getValidAndInvalidSegments().getValidSegments());
-        CarbonUpdateUtil.updateTableMetadataStatus(segmentSet, carbonTable, updateTime, true,
-            new ArrayList<String>());
+        List<String> segmentDeleteList;
+        if (segmentsToBeDeleted != null)
+        {
```
Comment (jackylk) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: move to previous line

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java [hunk 2]
```diff
@@ -98,12 +96,25 @@ public CarbonOutputCommitter(Path outputPath, TaskAttemptContext context) throws
...
           context.getConfiguration().get(CarbonTableOutputFormat.UPADTE_TIMESTAMP, null);
+      String segmentsToBeDeleted =
+          context.getConfiguration().get(CarbonTableOutputFormat.SEGMENTS_TO_BE_DELETED, null);
       if (updateTime != null) {
         Set<String> segmentSet = new HashSet<>(
             new SegmentStatusManager(carbonTable.getAbsoluteTableIdentifier())
                 .getValidAndInvalidSegments().getValidSegments());
-        CarbonUpdateUtil.updateTableMetadataStatus(segmentSet, carbonTable, updateTime, true,
-            new ArrayList<String>());
+        List<String> segmentDeleteList;
```
Comment (xuchuanyin) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: Let the default value of `SEGMENTS_TO_BE_DELETED` be `""` in line 100 will reduce the following code (line 105~111) to 1 line: Only line 108 is enough.

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java (hunk 1 above)
Comment (ravipesala) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: ok

File: hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java (hunk 2 above)
Comment (ravipesala) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/api/CarbonOutputCommitter.java]: Ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2525/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.0, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1301/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2693/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/DeleteExecution.scala [hunk 3]
```diff
@@ -58,14 +58,15 @@ object DeleteExecution {
       dataRdd: RDD[Row],
       timestamp: String,
       isUpdateOperation: Boolean,
-      executorErrors: ExecutionErrors): Boolean = {
+      executorErrors: ExecutionErrors): (Boolean, Seq[String]) = {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/DeleteExecution.scala]: It seems the first return value is always true
And it is hard to understand this function, can you add comment for it

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForDeleteCommand.scala [hunk 4]
```diff
@@ -77,7 +77,7 @@ private[sql] case class CarbonProjectForDeleteCommand(
         dataRdd,
         timestamp,
         isUpdateOperation = false,
-        executorErrors)) {
+        executorErrors)._1) {
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForDeleteCommand.scala]: It seems always true

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala [hunk 5]
```diff
@@ -107,11 +107,19 @@ with Serializable {
...
+      model.setFactTimeStamp(updateTimeStamp.get.toLong)
     }
+    // In case of update query there is chance to remove the older segments, so here we can set
+    // the to be deleted segments to mark as delete while updating tablestatus
+    val segemntsTobeDeleted = options.get("segmentsToBeDeleted")
+    if (segemntsTobeDeleted.isDefined) {
+      conf.set(CarbonTableOutputFormat.SEGMENTS_TO_BE_DELETED, segemntsTobeDeleted.get)
+    }
+
+
```
Comment (jackylk) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: remove empty line

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/DeleteExecution.scala (hunk 3 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/DeleteExecution.scala]: ok updated

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForDeleteCommand.scala (hunk 4 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/mutation/CarbonProjectForDeleteCommand.scala]: removed returning status

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala (hunk 5 above)
Comment (ravipesala) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/datasources/CarbonFileFormat.scala]: ok

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1321/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2559/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2727/

**Comment by ravipesala** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1337/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/2574/

**Comment by jackylk** [comment]
LGTM

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/2747/

**Pull request closed without merging**